modules/core/shared/src/main/scala/coursier/core/Dependency.scala

```diff
@@ -43,15 +47,73 @@ import dataclass.data
   def withPublication(name: String, `type`: Type, ext: Extension, classifier: Classifier): Dependency =
     withPublication(Publication(name, `type`, ext, classifier))
 
+  def withOptional(optional: Boolean) =  Dependency(module, version, configuration, exclusions, publication, optional, transitive)
+  def withTransitive(transitive: Boolean) =  Dependency(module, version, configuration, exclusions, publication, optional, transitive)
+
   lazy val clearExclusions: Dependency =
     withExclusions(Set.empty)
 
+  private val tuple = (this.module, this.version, this.configuration, this.exclusions, this.publication, this.optional, this.transitive)
   override lazy val hashCode: Int =
     tuple.hashCode()
+
+  override def toString: String = {
+    val b = new StringBuilder("Dependency(")
+    b.append(String.valueOf(module))
+    b.append(", ")
+    b.append(String.valueOf(version))
+    b.append(", ")
+    b.append(String.valueOf(configuration))
+    b.append(", ")
+    b.append(String.valueOf(exclusions))
+    b.append(", ")
+    b.append(String.valueOf(publication))
+    b.append(", ")
+    b.append(String.valueOf(optional))
+    b.append(", ")
+    b.append(String.valueOf(transitive))
+    b.append(")")
+    b.toString
+  }
 }
 
 object Dependency {
 
+  val memoised_cache: java.util.WeakHashMap[(Module, String, Configuration, Set[(Organization, ModuleName)], Publication, Boolean, Boolean), java.lang.ref.WeakReference[Dependency]] =
+    new java.util.WeakHashMap[(Module, String, Configuration, Set[(Organization, ModuleName)], Publication, Boolean, Boolean), java.lang.ref.WeakReference[Dependency]]()
+
+  def apply(module: Module, version: String, configuration: Configuration, exclusions: Set[(Organization, ModuleName)], publication: Publication, optional: Boolean, transitive: Boolean): Dependency = {
+    val key = (module, version, configuration, exclusions, publication, optional, transitive)
+    val first = {
+      val weak = memoised_cache.get(key)
+      if (weak == null) null else weak.get
+    }
+    if (first != null) {
+      first
+    } else {
+      memoised_cache.synchronized {
```

Comment:
I'd be wary of using `synchronized`… I would have just used a `java.util.concurrent.ConcurrentHashMap[(…), WeakReference[…]]` here, and handled concurrent insertions at the same key with `putIfAbsent`.

Reply:
Again used what stalagmite is using, but they have a warning about this: https://gitlab.com/fommil/attic/-/blob/master/stalagmite/src/test/scala/testing/weakmemoised/weakmemoised.scala#L69